Game.Checkers/CheckersState.cs

```diff
@@ -1,10 +1,159 @@
+        {
+            board = new Piece[BOARD_COLS, BOARD_ROWS];
+            Array.Copy(state.board, board, state.board.Length);
+            CurrentPlayer = state.CurrentPlayer;
+        }
         public bool Equals(CheckersState? other)
         {
-            throw new NotImplementedException();
+            if (other == null) return false;
+            if (CurrentPlayer != other.CurrentPlayer) return false;
+            for (int x = 0; x < BOARD_ROWS; x++)
+            {
+                for (int y = 0; y < BOARD_COLS; y++)
+                {
+                    if (GetPieceAt(x, y) != other.GetPieceAt(x, y)) return false;
+                }
+            }
+            return true;
+        }
+        public IEnumerable<Field> GetFields()
+        {
+            for (int col = 0; col < BOARD_ROWS; col++)
+            {
+                for (int row = col % 2; row < BOARD_COLS; row += 2)
+                {
+                    yield return new Field(col, row);
+                }
+            }
+        }
+        public IEnumerable<Field> GetNeighbours(Field field)
+        {
+            int x = field.Col;
+            int y = field.Row;
+            List<Field> neighbours = new();
+            if (y < BOARD_ROWS - 1)
+            {
+                if (x < BOARD_COLS - 1)
+                    neighbours.Add(new Field(x + 1, y + 1));
+                if (x > 0)
+                    neighbours.Add(new Field(x - 1, y + 1));
+            }
+            if (y > 0)
+            {
+                if (x < BOARD_COLS - 1)
+                    neighbours.Add(new Field(x + 1, y - 1));
+                if (x > 0)
+                    neighbours.Add(new Field(x - 1, y - 1));
+            }
+            return neighbours;
+        }
+
+        public Piece GetPieceAt(int col, int row)
+        {
+            return board[col, row];
+        }
+
+        public Piece GetPieceAt(Field field)
+        {
+            return GetPieceAt(field.Col, field.Row);
+        }
+
+        public void SetPieceAt(int col, int row, Piece piece)
+        {
+            board[col, row] = piece;
+        }
+        public void SetPieceAt(Field field, Piece piece)
+        {
+            SetPieceAt(field.Col, field.Row, piece);
+        }
+        public void SetPieceAt(Field field, Piece piece, bool possiblePromotion)
+        {
+            if (possiblePromotion)
+            {
+                if (piece == Piece.WhitePawn && field.Row == BOARD_ROWS - 1)
+                {
+                    SetPieceAt(field.Col, field.Row, Piece.WhiteCrowned);
+                    return;
+                }
+                if (piece == Piece.BlackPawn && field.Row == 0)
+                {
+                    SetPieceAt(field.Col, field.Row, Piece.BlackCrowned);
+                    return;
+                }
+            }
+            SetPieceAt(field.Col, field.Row, piece);
+        }
+        public void SetPieceAt(string field, Piece piece)
```

Comment:
czy poza testami ta funkcja jest gdziekolwiek używana? w ogóle po co sobie utrudniać życie w ten sposób?

Reply:
Jak to utrudniać? :O Właśnie bez tej nakładki myliłem się i pisałem bezsensowne testy, które potem nie działały. Funkcja jest  tylko pod testy.

Reply:
ok, nie wnikam xD